File: src/lib/gmp_lib/src/GMP/GMP_nDoF_IO.cpp
#include <gmp_lib/GMP/GMP_nDoF_IO.h>

namespace as64_
{

namespace gmp_
{

  GMP_nDoF_IO::GMP_nDoF_IO(const std::shared_ptr<gmp_::GMP_nDoF> &gmp)
  {
    this->gmp = gmp;
  }


  void GMP_nDoF_IO::write(const std::string &filename, const std::string &prefix)
  {
    gmp_::FileIO fid(filename, gmp_::FileIO::out|gmp_::FileIO::trunc);
    write(fid, prefix);
  }

  void GMP_nDoF_IO::write(gmp_::FileIO &fid, const std::string &prefix)
  {
    unsigned N_kernels = gmp->numOfKernels();
    unsigned n_dofs = gmp->numOfDoFs();

    fid.write(prefix + "weights", gmp->W);
    fid.write(prefix + "damping", gmp->D);
    fid.write(prefix + "stiffness", gmp->K);
    fid.write(prefix + "N_kernels", N_kernels);
    fid.write(prefix + "N_DoFs", n_dofs);
    fid.write(prefix + "scale_type", gmp->traj_sc->getScaleType());
    fid.write(prefix + "c", gmp->c);
    fid.write(prefix + "h", gmp->h);
  }

  void GMP_nDoF_IO::read(const std::string &filename, const std::string &prefix)
  {
    gmp_::FileIO fid(filename, gmp_::FileIO::in);
    read(fid, prefix);
  }

  void GMP_nDoF_IO::read(gmp_::FileIO &fid, const std::string &prefix)
  {
    fid.read(prefix + "weights", gmp->W);
    fid.read(prefix + "damping", gmp->D);
    fid.read(prefix + "stiffness", gmp->K);
    //fid.read(prefix + "N_kernels", N_kernels);
    //fid.read(prefix + "N_DoFs", n_dofs);
    int scale_type;
    fid.read(prefix + "scale_type", scale_type);
    fid.read(prefix + "c", gmp->c);
    fid.read(prefix + "h", gmp->h);

    gmp->Y0d = gmp->W*gmp->regressVec(0);
    gmp->Ygd = gmp->W*gmp->regressVec(1);

    gmp->setY0(gmp->Y0d);
    gmp->setGoal(gmp->Ygd);

    unsigned n_dofs = gmp->numOfDoFs();
    gmp->y_dot = arma::vec().zeros(n_dofs);
    gmp->z_dot = arma::vec().zeros(n_dofs);

    if (scale_type == TrajScale::PROP_SCALE) gmp->setScaleMethod( gmp_::TrajScale::Ptr(new gmp_::TrajScale_Prop(n_dofs)) );
    else if (scale_type == TrajScale::ROT_MIN_SCALE) gmp->setScaleMethod( gmp_::TrajScale::Ptr(new gmp_::TrajScale_Rot_min()) );
    else if (scale_type == TrajScale::ROT_WB_SCALE) gmp->setScaleMethod( gmp_::TrajScale::Ptr(new gmp_::TrajScale_Rot_wb()) );
    else throw std::runtime_error("[GMP_nDoF_IO::read]: Unsupported scale type \"" + std::to_string(scale_type) + "\"...\n");
  }


} // namespace gmp_

} // namespace as64_

File: src/gmp_test/matlab/lib/gmp_lib/GMP_nDoF/GMP_nDoF_IO.m
%% N-DoF GMP class
%  Generalized movement primitive.
%

classdef GMP_nDoF_IO < matlab.mixin.Copyable
    
    methods (Static, Access = public)

        %% Write the GMP model to a file.
        % @param[in] gmp: Pointer to a @GMP_nDoF object.
        % @param[in] fid: Filename string or object of type @FileIO associated with the file.
        % @param[in] prefix: The prefix that will be used for writing the names of all GMP params (optional, default="").
        function write(fid, prefix)
            
            if (nargin < 3), prefix=''; end
            
            if (ischar(fid))
               filename = fid;
               fid = FileIO(filename, bitor(FileIO.out,FileIO.trunc));
            end
            
            N_kernels = uint32(gmp.numOfKernels());
            n_dofs = uint32(gmp.numOfDoFs());
            
            fid.write([prefix 'weights'], gmp.W);
            fid.write([prefix 'damping'], gmp.D);
            fid.write([prefix 'stiffness'], gmp.K);
            fid.write([prefix 'N_kernels'], N_kernels);
            fid.write([prefix 'N_DoFs'], n_dofs);
            fid.write([prefix 'scale_type'], gmp.traj_sc.getScaleType());
            fid.write([prefix 'c'], gmp.c);
            fid.write([prefix 'h'], gmp.h);
            
        end
        
        %% Reads the GMP model from a file.
        % @param[in] gmp: Pointer to a @GMP_nDoF object.
        % @param[in] fid: Filename string or object of type @FileIO associated with the file.
        % @param[in] prefix: The prefix that will be used for reading the names of all GMP params (optional, default="").
        function read(fid, prefix)
            
            if (nargin < 3), prefix=''; end
            
            if (ischar(fid))
               filename = fid;
               fid = FileIO(filename, FileIO.in);
            end
            
            gmp.W = fid.read([prefix 'weights']);
            gmp.D = fid.read([prefix 'damping']);
            gmp.K = fid.read([prefix 'stiffness']);
            %N_kernels = fid.read([prefix 'N_kernels']);
            %n_dofs = fid.read([prefix 'N_DoFs']);
            scale_type = fid.read([prefix 'scale_type']);
            gmp.c = fid.read([prefix 'c']);
            gmp.h = fid.read([prefix 'h']);
            
            n_dofs = gmp.numOfDoFs();
            
            gmp.Y0d = gmp.W*gmp.regressVec(0);
            gmp.Ygd = gmp.W*gmp.regressVec(1);
            gmp.Y0 = gmp.Y0d;
            gmp.Yg = gmp.Ygd;
            gmp.y_dot = zeros(n_dofs,1);
            gmp.z_dot = zeros(n_dofs,1);
            
            if (scale_type == TrajScale.PROP_SCALE), gmp.setScaleMethod( TrajScale_Prop(n_dofs) );
            elseif (scale_type == TrajScale.ROT_MIN_SCALE), gmp.setScaleMethod( TrajScale_Rot_min() );
            elseif (scale_type == TrajScale.ROT_WB_SCALE), gmp.setScaleMethod( TrajScale_Rot_wb() );
            else, error(['[GMP_nDoF_IO::read]: Unsupported scale type ''' num2str(scale_type) '''...\n']);
            end

    end
    
end
